src/mpid/ch4/generic/gpu/mpidig_gpu_utils.h

```diff
@@ -0,0 +1,183 @@
+{
+    int dt_contig;
+    size_t size;
+    MPIR_Datatype *dt_ptr;
+    MPI_Aint true_lb;
+    yaksa_request_t request;
+    uintptr_t actual_pack_bytes;
+    int rc;
+
+    MPIDI_Datatype_get_info(count, datatype, dt_contig, size, dt_ptr, true_lb);
+    if (dt_contig) {
+        /* Stage buffer is contiguous, we can move directly. */
+        rc = yaksa_ipack(sbuf, (uintptr_t) size, YAKSA_TYPE__BYTE, 0,
+                         dbuf, (uintptr_t) size, &actual_pack_bytes, &request);
+        MPIR_Assert(rc == YAKSA_SUCCESS);
+        rc = yaksa_request_wait(request);
+        MPIR_Assert(rc == YAKSA_SUCCESS);
+    } else {
+        double density = ((double) (dt_ptr->size)) / ((double) (dt_ptr->extent));
+        if (density >= MPIDIG_GPU_DT_HIGH_DENSITY) {
+            /* No need to pack for high-density datatype */
+            /* We convert it to iov, then copy each iov into stage buffer */
+            int i;
+            MPI_Aint curr_len;
+            void *curr_loc;
+            MPI_Aint num_iov;
+            int n_iov;
+            MPL_IOV *dt_iov;
+            int actual_iov_len;
+            MPI_Aint actual_iov_bytes;
+            size_t accumulated_count = 0;
+
+            MPIR_Typerep_iov_len(count, datatype, 0, size, &num_iov);
+            n_iov = (int) num_iov;
+            dt_iov = (MPL_IOV *) MPL_malloc(n_iov * sizeof(MPL_IOV), MPL_MEM_BUFFER);
+            MPIR_Typerep_to_iov(NULL, count, datatype, 0, dt_iov, n_iov, size, &actual_iov_len,
+                                &actual_iov_bytes);
+            n_iov = actual_iov_len;
+            MPIR_Assert(actual_iov_bytes == (MPI_Aint) size);
+
+            for (i = 0; i < n_iov; i++) {
+                curr_loc = dt_iov[i].MPL_IOV_BUF;
+                curr_len = dt_iov[i].MPL_IOV_LEN;
+                rc = yaksa_ipack((char *) sbuf + MPIR_Ptr_to_aint(curr_loc), (uintptr_t) curr_len,
+                                 YAKSA_TYPE__BYTE, 0, (char *) dbuf + accumulated_count,
+                                 (uintptr_t) curr_len, &actual_pack_bytes, &request);
+                MPIR_Assert(rc == YAKSA_SUCCESS);
+                rc = yaksa_request_wait(request);
+                MPIR_Assert(rc == YAKSA_SUCCESS);
+                accumulated_count += curr_len;
+            }
+            MPL_free(dt_iov);
+        } else {
+            /* We need to pack for low-density datatype */
+            /* Allocate temporary buffer, which includes whole exent for datatype */
+            size_t tmp_sz = dt_ptr->extent * count;
+            void *tmp_buf = NULL;
+            MPL_gpu_malloc_host(&tmp_buf, tmp_sz);
+            /* Move whole non-contiguous data from GPU to temporary buffer. */
+            rc = yaksa_ipack(sbuf, (uintptr_t) tmp_sz, YAKSA_TYPE__BYTE, 0,
+                             tmp_buf, (uintptr_t) tmp_sz, &actual_pack_bytes, &request);
+            MPIR_Assert(rc == YAKSA_SUCCESS);
+            rc = yaksa_request_wait(request);
+            MPIR_Assert(rc == YAKSA_SUCCESS);
```

Comment:
Please add a FIXME over here because this is only a temporary hack till yaksa gets integrated into typerep.